sd: disk ioctl get card cid command #66460

tuna-f1sh opened the pull request:
I also wanted access to card CID (#66460) so added a disk ioctl command to retrieve this - cached in `sd_card struct` rather than stack in `card_read_cid`.

Some option questions/opinions:

* Should I re-order and re-number commands so it appears before INIT/DEINIT? Just tidier before CTRL commands I guess.
* The `struct sd_cid` could be behind the `CONFIG_SDMMC_STACK`.. guard since it's only used in this case.
* Even cleaner, could just return cid_raw since it's read in all configurations (or another command to get this)? I have to admit, I don't fully understand why the raw value is always read but only decoded with SDMMC and otherwise discarded.

tuna-f1sh committed:
sd: disk ioctl get card cid command #66460

CID (card id register) can be useful to application for traceability of SD
cards.

Adds disk ioctl command to get the CID data that is read at initialisation.
The struct sd_cid is moved to sd_card so that it can be read and retrieved
by card_ioctrl.

Signed-off-by: John Whittington <git@jbrengineering.co.uk>

decsny commented:
what is the use case where you need the card ID?

danieldegrasse reviewed (commented):
I don't feel too strongly about this, but the CID was originally not placed in the SD card structure to reduce RAM utilization. I'd rather we didn't cache it, and instead read it from the card when an IOCTL was issued.

tuna-f1sh committed:
sd: disk ioctl get card cid command

CID (card id register) can be useful to application for traceability of SD
cards (#66460).

Adds disk ioctl command to get the CID data that is read at initialisation.

Changes the card_read_cid function to take pointer to cid register so that
the ioctl command can utilize with the passed user buf. Updated
calling functions to pass this pointer.

Signed-off-by: John Whittington <git@jbrengineering.co.uk>

tuna-f1sh commented:
Thanks. I've pushed another commit considering this:

* Removed the cached decoded CID struct.
* Changed `card_read_cid` to take pointer to CID register to fill. Allows the IOCTL function to use it alongside the initialisation code.
* To keep things simple, changed the IOCTL command so that it files the raw CID register to the passed `(void*) buf` rather than doing the decode too. It prevents having to create a `uint32_t cid[4]` on stack in the IOCTL function and I figured a user reading this should probably know what the CID register contains and do their own decoding for the application's specific usage.

This means there should be no changes to RAM usage. Could expose the `sdmmc_decode_cid` rather than it being static inline but probably better to leave it to the user application.

kartben commented on include/zephyr/drivers/disk.h:
```diff
@@ -68,6 +68,11 @@ extern "C" {
  * requested, but this operation is inherently unsafe.
  */
 #define DISK_IOCTL_CTRL_DEINIT			7
+/* Get Card identification (CID) register value.
```
```suggestion
/** Get Card identification (CID) register value.
```

tuna-f1sh replied:
Thanks!

FacepalmMute commented:
@tuna-f1sh thank you for your patch! Backported this to my v4.2 zephyr fork.
Was also wondering why the CID is discarded by zephyr.
